### PR TITLE
Admin tool testing

### PR DESCRIPTION
**Pull request by mattcollier**

### CODE REVIEW

File: lib/services.identity.js [hunk 1]
```diff
@@ -162,6 +175,18 @@ bedrock.events.on('bedrock-express.configure.routes', function(app) {
     ensureAuthenticated,
     validate({query: 'services.identity.getIdentitiesQuery'}),
     function(req, res, next) {
+      // TODO: support URL query parameters
+      var q = {
+        'identity.sysStatus': 'active'
+      };
+      brIdentity.getAll(null, q, function(err, records) {
+        if (err) {
```
Comment (dlongley) [inline lib/services.identity.js]: Remove extra space.

File: lib/services.identity.js [hunk 2]
```diff
@@ -376,6 +375,61 @@ bedrock.events.on('bedrock-express.configure.routes', function(app) {
...
+              var identity = operation.identity;
+              brIdentity.setRoles(null, identity, function(err, record) {
+                return callback(err, record);
+              });
+            }, function(err, results) {
+              return callback(err, results);
+            });
+        }
+      }, function(err, results) {
+          if(err) {
```
Comment (dlongley) [inline lib/services.identity.js]: Fix indent.

File: lib/services.identity.js [hunk 3]
```diff
@@ -162,6 +169,18 @@ bedrock.events.on('bedrock-express.configure.routes', function(app) {
     ensureAuthenticated,
     validate({query: 'services.identity.getIdentitiesQuery'}),
     function(req, res, next) {
+      // TODO: support URL query parameters
+      var q = {
+        'identity.sysStatus': 'active'
+      };
+      brIdentity.getAll(null, q, function(err, records) {
```
Comment (dlongley) [inline lib/services.identity.js]: This call has no authentication check. Anyone can hit this endpoint and get all of the active identities stored in the database. Change `null` to `req.user.identity`.

File: lib/services.identity.js [hunk 4]
```diff
@@ -162,6 +169,18 @@ bedrock.events.on('bedrock-express.configure.routes', function(app) {
     ensureAuthenticated,
     validate({query: 'services.identity.getIdentitiesQuery'}),
     function(req, res, next) {
```
Comment (dlongley) [inline lib/services.identity.js]: This API call has documentation below it that does not apply at all. This should be commented out or updated in some way.

File: lib/services.identity.js [hunk 5]
```diff
@@ -210,68 +230,90 @@ bedrock.events.on('bedrock-express.configure.routes', function(app) {
...
         }
         res.sendStatus(204);
       });
     });
+  docs.annotate.patch(idPath, {
+    description: 'Update an existing identity on the system.',
+    securedBy: ['cookie', 'hs1'],
+    schema: 'services.identity.patchIdentity',
+    responses: {
+      204: 'The identity was updates successfully.',
```
Comment (dlongley) [inline lib/services.identity.js]: Typo.

File: lib/services.identity.js [hunk 6]
```diff
@@ -292,38 +334,52 @@ bedrock.events.on('bedrock-express.configure.routes', function(app) {
...
       getId: getIdentityIdFromUrl.bind(null, req),
+      checkPermission: ['getId', function(callback, results) {
+        if(req.isAuthenticated() && 'identity' in req.user) {
+          return brPermission.checkPermission(
+            req.user.identity, PERMISSIONS.IDENTITY_ACCESS,
+            {resource: results.getId}, function(err, result) {
+              if(err) {
+                return callback(null, false);
+              }
+              return callback(null, true);
```
Comment (dlongley) [inline lib/services.identity.js]: Simplify to: `return callback(null, !err);`

File: lib/services.identity.js [hunk 7]
```diff
@@ -376,6 +432,60 @@ bedrock.events.on('bedrock-express.configure.routes', function(app) {
       404: 'The identity was not found on the system.'
     }
   });
+
+  app.delete(idPath, ensureAuthenticated, function(req, res, next) {
+    var fullUrl = req.protocol + '://' + req.get('host') + req.originalUrl;
```
Comment (dlongley) [inline lib/services.identity.js]: This doesn't look like the right way to do this. cc: @davidlehn -- what's the right fix here?

File: lib/services.identity.js [hunk 8]
```diff
@@ -376,6 +432,60 @@ bedrock.events.on('bedrock-express.configure.routes', function(app) {
       404: 'The identity was not found on the system.'
     }
   });
+
+  app.delete(idPath, ensureAuthenticated, function(req, res, next) {
```
Comment (dlongley) [inline lib/services.identity.js:429]: @davidlehn -- are we comfortable with exposing deletion of identities (marking them deleted)?

File: lib/services.identity.js [hunk 9]
```diff
@@ -376,6 +432,60 @@ bedrock.events.on('bedrock-express.configure.routes', function(app) {
...
+              });
+            }, function(err, results) {
+              return callback(err, results);
+            });
+        }
+      }, function(err, results) {
+        if(err) {
+          return next(err);
+        }
+        res.json(results);
```
Comment (dlongley) [inline lib/services.identity.js]: We should not be returning `results` from the async API to the client. Maybe a 204 instead?

File: lib/services.identity.js [hunk 10]
```diff
@@ -376,6 +432,60 @@ bedrock.events.on('bedrock-express.configure.routes', function(app) {
...
+            return callback();
+          }
+          async.each(updateRoleOperations,
+            function(operation, callback) {
+              var identity = operation.identity;
+              brIdentity.setRoles(null, identity, function(err, record) {
+                return callback(err, record);
+              });
+            }, function(err, results) {
+              return callback(err, results);
```
Comment (dlongley) [inline lib/services.identity.js]: Why are we passing on `results`?

File: lib/services.identity.js [hunk 11]
```diff
@@ -376,6 +432,60 @@ bedrock.events.on('bedrock-express.configure.routes', function(app) {
...
+        },
+        updateRoles: function(callback) {
+          if(updateRoleOperations.length === 0) {
+            return callback();
+          }
+          async.each(updateRoleOperations,
+            function(operation, callback) {
+              var identity = operation.identity;
+              brIdentity.setRoles(null, identity, function(err, record) {
+                return callback(err, record);
```
Comment (dlongley) [inline lib/services.identity.js]: Why are we passing `record`?

File: lib/services.identity.js [hunk 12]
```diff
@@ -376,6 +432,60 @@ bedrock.events.on('bedrock-express.configure.routes', function(app) {
...
+      var updateRoleOperations = operations.filter(function(operation) {
+        return operation.op === 'updateRoles';
+      });
+
+      async.auto({
+        delete: function(callback) {
+          return callback();
+        },
+        update: function(callback) {
+          return callback();
```
Comment (dlongley) [inline lib/services.identity.js]: This should produce a not implemented error.

File: lib/services.identity.js [hunk 13]
```diff
@@ -376,6 +432,60 @@ bedrock.events.on('bedrock-express.configure.routes', function(app) {
...
+      var updateOperations = operations.filter(function(operation) {
+        return operation.op === 'update';
+      });
+      var updateRoleOperations = operations.filter(function(operation) {
+        return operation.op === 'updateRoles';
+      });
+
+      async.auto({
+        delete: function(callback) {
+          return callback();
```
Comment (dlongley) [inline lib/services.identity.js]: This should produce a not implemented error.

File: lib/services.identity.js (hunk 7 above)
Comment (davidlehn) [inline lib/services.identity.js]: Something like:
var identityId = createIdentityId(req.params.identity);

File: lib/services.identity.js (hunk 8 above)
Comment (davidlehn) [inline lib/services.identity.js:429]: I'm not sure what else that impacts right now.  If an id is just marked deleted, do the admin queries still show it?  Can an admin undelete it?  Can a id owner delete themselves?  Do all the other identity queries handle that an id could be marked deleted?

Comment (dlongley) [inline lib/services.identity.js:429]: You do need to have the IDENTITY_ADMIN permission to change identity status to deleted.

File: lib/services.identity.js [hunk 14]
```diff
@@ -140,7 +147,7 @@ bedrock.events.on('bedrock-express.configure.routes', function(app) {
         }
         // return identity
         res.set('Location', identity.id);
```
Comment (davidlehn) [inline lib/services.identity.js]: Use `result.identity.id` to ensure synced with return data below.

File: lib/services.identity.js (hunk 8 above)
Comment (mattcollier) [inline lib/services.identity.js:429]: @dlongley @davidlehn Presently an identity that is not 'active' is not visible to the admin interface.  Do we need to have undelete functions in the admin UI?

We do need to disallow deletion of ones own identity in the UI.

**Pull request merged**